File: go/core/transport.go
package core

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"github.com/royaltics/tracker-go/types"
)

type Transport struct {
	config *types.ClientConfig
	client *http.Client
}

func NewTransport(config *types.ClientConfig) *Transport {
	return &Transport{
		config: config,
		client: &http.Client{
			Timeout: config.Timeout,
		},
	}
}

func (t *Transport) Send(compressedEvent string) error {
	payload := types.TransportPayload{
		Event:         compressedEvent,
		LicenseID:     t.config.LicenseID,
		LicenseName:   t.config.LicenseName,
		LicenseDevice: t.config.LicenseDevice,
	}

	return t.makeRequest(payload)
}

func (t *Transport) makeRequest(payload types.TransportPayload) error {
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("failed to marshal payload: %w", err)
	}

	req, err := http.NewRequest("POST", t.config.WebhookURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "Royaltics-ErrorTracker-Go/1.0")

	for key, value := range t.config.Headers {
		req.Header.Set(key, value)
	}

	resp, err := t.client.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	io.Copy(io.Discard, resp.Body)

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, resp.Status)
	}

	return nil
}

File: php/src/Core/Transport.php
<?php

declare(strict_types=1);

namespace Royaltics\ErrorTracker\Core;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use Royaltics\ErrorTracker\Types\ClientConfig;
use Royaltics\ErrorTracker\Types\TransportPayload;

final class Transport
{
    private readonly Client $client;

    public function __construct(private readonly ClientConfig $config)
    {
        $this->client = new Client([
            'timeout' => $this->config->timeout / 1000,
            'connect_timeout' => $this->config->timeout / 1000,
        ]);
    }

    public function send(string $compressedEvent): void
    {
        $payload = new TransportPayload(
            event: $compressedEvent,
            licenseId: $this->config->licenseId,
            licenseDevice: $this->config->licenseDevice,
            licenseName: $this->config->licenseName
        );

        $lastError = null;

        for ($attempt = 0; $attempt <= $this->config->maxRetries; $attempt++) {
            try {
                $this->makeRequest($payload);
                return;
            } catch (\Exception $e) {
                $lastError = $e;

                if ($attempt < $this->config->maxRetries) {
                    usleep($this->calculateBackoff($attempt) * 1000);
                }
            }
        }

        throw $lastError ?? new \RuntimeException('Transport failed with unknown error');
    }

    private function makeRequest(TransportPayload $payload): void
    {
        $jsonPayload = json_encode([
            'event' => $payload->event,
            'license_id' => $payload->licenseId,
            'license_name' => $payload->licenseName,
            'license_device' => $payload->licenseDevice,
        ], JSON_THROW_ON_ERROR);

        $headers = array_merge(
            [
                'Content-Type' => 'application/json',
                'User-Agent' => 'Royaltics-ErrorTracker-PHP/1.0',
            ],
            $this->config->headers
        );

        try {
            $response = $this->client->post($this->config->webhookUrl, [
                'body' => $jsonPayload,
                'headers' => $headers,
            ]);

            if ($response->getStatusCode() < 200 || $response->getStatusCode() >= 300) {
                throw new \RuntimeException(
                    sprintf('HTTP %d: %s', $response->getStatusCode(), $response->getReasonPhrase())
                );
            }
        } catch (GuzzleException $e) {
            throw new \RuntimeException('Network error: ' . $e->getMessage(), 0, $e);
        }
    }

    private function calculateBackoff(int $attempt): int
    {
        $baseDelay = 1000;
        $maxDelay = 30000;
        $delay = min($baseDelay * (2 ** $attempt), $maxDelay);
        
        return $delay;
    }
}
